### PR TITLE
Add alias_cert model check when set cert

### PR DESCRIPTION
**Pull request by Wenxing-hou**
Fix: #2066

**Commit 1:** Add alias_cert model check in set_cert_rsp;
**Commit 2:** Fix unit_test after adding alias_cert model check;
**Commit 3:** Generate alias_cert model cert_chain for responder;
**Commit 4:** Enhance unit_test;
**Commit 5:** Add `LIBSPDM_CERT_PARSE_SUPPORT` for set_cert ;

### CODE REVIEW

File: library/spdm_crypt_lib/libspdm_crypt_cert.c [hunk 1]
```diff
@@ -88,6 +88,18 @@
 #define ENCRY_ALGO_ED448_OID {0x2B, 0x65, 0x71}
 #endif
 
+/*leaf cert basic_constraints false case1: CA: false and CA object is excluded */
+#define BASIC_CONSTRAINTS_STRING_FALSE_CASE1 {0x30, 0x00}
+uint8_t basic_constraints_false_case1[] = BASIC_CONSTRAINTS_STRING_FALSE_CASE1;
```
Comment (jyao1) [inline library/spdm_crypt_lib/libspdm_crypt_cert.c:93]: please dont use global variable

File: os_stub/spdm_device_secret_lib_sample/cert.c [hunk 2]
```diff
@@ -496,6 +496,152 @@ bool libspdm_read_responder_public_certificate_chain(
     return true;
 }
 
+bool libspdm_read_responder_public_certificate_chain_alias_cert(
```
Comment (jyao1) [inline os_stub/spdm_device_secret_lib_sample/cert.c:499]: This name is confusing.

Is this full alias cert chain till leaf? 
or alias cert chain till device cert CA?

File: unit_test/sample_key/readme.txt [hunk 3]
```diff
@@ -618,3 +618,52 @@ openssl asn1parse -in end_responder_with_spdm_req_rsp_eku.cert -out end_responde
 openssl asn1parse -in end_responder_with_spdm_req_eku.cert -out end_responder_with_spdm_req_eku.cert.der
 openssl asn1parse -in end_responder_with_spdm_rsp_eku.cert -out end_responder_with_spdm_rsp_eku.cert.der
 popd
+
+==== More alias_cert model cert_chain to gen ====
+NOTE: The bundle_responder.certchain_alias.der and bundle_requester.certchain.der have same ca_cert and inter cert.
```
Comment (jyao1) [inline unit_test/sample_key/readme.txt:623]: Please add:
This alias cert chain is partial, from root CA to device certificate CA.

File: library/spdm_crypt_lib/libspdm_crypt_cert.c (hunk 1 above)
Comment (Wenxing-hou) [reply library/spdm_crypt_lib/libspdm_crypt_cert.c]: Thanks. I have changed the code.

File: unit_test/sample_key/readme.txt (hunk 3 above)
Comment (Wenxing-hou) [reply unit_test/sample_key/readme.txt:623]: Thanks. I have added the comment.

File: library/spdm_crypt_lib/libspdm_crypt_cert.c [hunk 4]
```diff
@@ -1052,6 +1110,73 @@ bool libspdm_x509_certificate_check(const uint8_t *cert, size_t cert_size,
...
 
+/**
+ * Certificate Check for SPDM leaf cert when get_cert command
+ *
+ * @param[in]  cert                  Pointer to the DER-encoded certificate data.
+ * @param[in]  cert_size             The size of certificate data in bytes.
+ * @param[in]  base_asym_algo        SPDM base_asym_algo
+ * @param[in]  base_hash_algo        SPDM base_hash_algo
+ * @param[in]  is_requester_cert     Is the function verifying requester or responder cert.
+ * @param[in]  is_device_cert_model  If true, the cert chain is DeviceCert model;
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/libspdm_crypt_cert.c:1121]:
```
If true, the local endpoint uses the DeviceCert model.
If false, the local endpoint uses the AliasCert model.
```

**Review by steven-bellock — CHANGES_REQUESTED** [review]
Most of this should be gated by `LIBSPDM_CERT_PARSE_SUPPORT`. This brings in a lot of code just for checking the certificate chain, and a Responder implementation may be OK with accepting the certificate chain as-is.

File: include/library/spdm_crypt_lib.h [hunk 5]
```diff
@@ -825,6 +825,24 @@ bool libspdm_x509_certificate_check(const uint8_t *cert, size_t cert_size,
...
 
+/**
+ * Certificate Check for SPDM leaf cert when set_cert.
+ *
+ * @param[in]  cert                  Pointer to the DER-encoded certificate data.
+ * @param[in]  cert_size             The size of certificate data in bytes.
+ * @param[in]  base_asym_algo        SPDM base_asym_algo
+ * @param[in]  base_hash_algo        SPDM base_hash_algo
+ * @param[in]  is_requester_cert     Is the function verifying requester or responder cert.
+ * @param[in]  is_device_cert_model  If true, the cert chain is DeviceCert model.
```
[the remark this answers is not in the document]
Comment (Wenxing-hou) [reply include/library/spdm_crypt_lib.h]: Thanks. I have changed the comment.

**Comment by Wenxing-hou** [comment]
> Most of this should be gated by `LIBSPDM_CERT_PARSE_SUPPORT`. This brings in a lot of code just for checking the certificate chain, and a Responder implementation may be OK with accepting the certificate chain as-is.

Thanks. I have added the `LIBSPDM_CERT_PARSE_SUPPORT` for set_cert.

File: library/spdm_crypt_lib/libspdm_crypt_cert.c [hunk 6]
```diff
@@ -762,23 +766,85 @@ static bool libspdm_verify_leaf_cert_basic_constraints(const uint8_t *cert, size
...
+    uint8_t basic_constraints_true_case[] = BASIC_CONSTRAINTS_STRING_TRUE_CASE;
+
+    len = LIBSPDM_MAX_BASIC_CONSTRAINTS_CA_LEN;
+
+    status = libspdm_x509_get_extended_basic_constraints(cert, cert_size,
+                                                         cert_basic_constraints, &len);
+
+    if (is_device_cert_model) {
+        /*device cert model*/
+        if (len == 0) {
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/libspdm_crypt_cert.c:815]: `status` should be checked before `len` is evaluated.

File: library/spdm_crypt_lib/libspdm_crypt_cert.c [hunk 7]
```diff
@@ -762,23 +766,85 @@ static bool libspdm_verify_leaf_cert_basic_constraints(const uint8_t *cert, size
...
+
+        if ((len == sizeof(basic_constraints_false_case2)) &&
+            (libspdm_consttime_is_mem_equal(cert_basic_constraints,
+                                            basic_constraints_false_case2,
+                                            sizeof(basic_constraints_false_case2)))) {
+            return true;
+        }
+    } else {
+        /*alias cert model*/
+        if ((len == sizeof(basic_constraints_true_case)) &&
```
Comment (steven-bellock) [inline library/spdm_crypt_lib/libspdm_crypt_cert.c:837]: `status` needs to be checked.

Comment (Wenxing-hou) [reply library/spdm_crypt_lib/libspdm_crypt_cert.c]: Thanks. I have added the check.

**Pull request merged**